### PR TITLE
Fix pillbox not uncloaking upon firing

### LINKED ISSUE #21093: AttackedGarrisoned does not trigger uncloak
## Issue Summary

This issue is for bleed. Actors with the AttackGarrisoned trait no longer uncloak with this attack. It appears to stem from https://github.com/OpenRA/OpenRA/pull/20675/commits/fca63d79cc65dfa18d43aee0dc8f8eba3fc35229

## Reproduction

Build a Camo Pillbox in Red Alert, force-attack the ground, and see if the structure uncloaks.

I've also thrown in a test map. This has some cloaked pillboxes, a normal Phase Transport, a Phase Transport with a garrison attack, and a cloaked Turret for comparison.
[ghostbox.zip](https://github.com/OpenRA/OpenRA/files/12818205/ghostbox.zip)

## Expected behavior

A Camo Pillbox should temporarily uncloak when it attacks.

## Screen recordings

Some pillboxes that are visible for the first few moments, before their cloak starts. A cloaked Turret acts as expected:

https://github.com/OpenRA/OpenRA/assets/4985264/87c5d4d4-3877-45d4-a2be-3e048527fc8a

Some pillboxes are detected and return fire, but they remain cloaked:

https://github.com/OpenRA/OpenRA/assets/4985264/51e72003-50af-4848-a516-450889bd5b7f

### PR DESCRIPTION
**Pull request by PunkPun**
Regression from #20675
Fixes #21093

### CODE REVIEW

**Comment by JovialFeline** [comment]
Fixes the tests I had, including the odd multi-passenger pillbox on wheels. 👌

**Comment by dnqbob** [comment]
Call other actor's `INotifyAttack` will result in messing up traits like `GrantConditionOnAttack` and crashing.

```
Exception of type `System.InvalidOperationException`: Attempting to revoke condition with invalid token 2 for hbox 292.
   at OpenRA.Actor.RevokeCondition(Int32 token) in D:\work\Projects\Github\OpenRA\OpenRA.Game\Actor.cs:line 583
   at OpenRA.Mods.Common.Traits.GrantConditionOnAttack.RevokeInstance(Actor self, Boolean revokeAll) in D:\work\Projects\Github\OpenRA\OpenRA.Mods.Common\Traits\Conditions\GrantConditionOnAttack.cs:line 77
   at OpenRA.TraitDictionary.TraitContainer`1.ApplyToAllTimed(Action`2 action, String text) in D:\work\Projects\Github\OpenRA\OpenRA.Game\TraitDictionary.cs:line 312
   at OpenRA.World.Tick() in D:\work\Projects\Github\OpenRA\OpenRA.Game\World.cs:line 458
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in D:\work\Projects\Github\OpenRA\OpenRA.Game\Game.cs:line 646
   at OpenRA.Game.LogicTick() in D:\work\Projects\Github\OpenRA\OpenRA.Game\Game.cs:line 661
   at OpenRA.Game.Loop() in D:\work\Projects\Github\OpenRA\OpenRA.Game\Game.cs:line 830
   at OpenRA.Game.Run() in D:\work\Projects\Github\OpenRA\OpenRA.Game\Game.cs:line 883
   at OpenRA.Game.InitializeAndRun(String[] args) in D:\work\Projects\Github\OpenRA\OpenRA.Game\Game.cs:line 313
   at OpenRA.Launcher.Program.Main(String[] args) in D:\work\Projects\Github\OpenRA\OpenRA.Launcher\Program.cs:line 32
```

Testcase: replace the same file(yaml) in OpenRA-RA 
[structures.zip](https://github.com/OpenRA/OpenRA/files/13445420/structures.zip)

**Comment by dnqbob** [comment]
I really don't recommend use this way to fix the uncloak issue because there can be countless dangerous edge cases on traits like `GrantConditionOnAttack` (for example you also need to test cases when `IsCyclic` is enabled and applying proper and hacky fix after that).

**Comment by PunkPun** [comment]
Fixed the crash

**Comment by PunkPun** [comment]
I think it makes perfect sense for infantry to call the pillbox's NotifyAttack. Not having that created weird exceptions and limitations to the AttackGarrison trait.

**Comment by dnqbob** [comment]
Then I recommend a detailed test on all INotifiedAttack related traits. They can have flaws on function but should never crash.

**Comment by PunkPun** [comment]
This fixes traits like ammo and reveal on fire not working in pillboxes

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -184,6 +184,16 @@ protected override void Created(Actor self)
 			base.Created(self);
 		}
 
+		public void AddNotifyAttacks(Actor attacker, INotifyAttack[] notifyAttacks)
+		{
+			this.notifyAttacks = this.notifyAttacks.Append(notifyAttacks.Select(a => (attacker, a)).ToArray()).ToArray();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Armament.cs]: That shouldn't be needed.

```suggestion
			this.notifyAttacks = this.notifyAttacks.Append(notifyAttacks.Select(a => (attacker, a))).ToArray();
```

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 2]
```diff
@@ -347,11 +353,11 @@ protected virtual void FireBarrel(Actor self, IFacing facing, in Target target,
...
 						Game.Sound.Play(SoundType.World, args.Weapon.StartBurstReport, self.World, self.CenterPosition);
 
-					foreach (var na in notifyAttacks)
-						na.Attacking(self, delayedTarget, this, barrel);
-
 					Recoil = Info.Recoil;
 				}
+
+				foreach (var (notifyActor, notify) in notifyAttacks)
+					notify.Attacking(notifyActor, delayedTarget, this, barrel);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Armament.cs:360]: Why is this moved outside of the delayed action? Won't that break if there is a delay?

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 3]
```diff
@@ -118,7 +118,7 @@ public class Armament : PausableConditionalTrait<ArmamentInfo>, ITick
 
 		BodyOrientation coords;
 		INotifyBurstComplete[] notifyBurstComplete;
-		INotifyAttack[] notifyAttacks;
+		(Actor NotifyActor, INotifyAttack Notify)[] notifyAttacks;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Why not make this a list instead of an array? That would save the gymnastics when adding or removing.

File: OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs [hunk 4]
```diff
@@ -164,7 +185,7 @@ public override void DoAttack(Actor self, in Target target)
 				paxFacing[a.Actor].Facing = targetYaw;
 				paxPos[a.Actor].SetCenterPosition(a.Actor, pos + PortOffset(self, port));
 
-				if (!a.CheckFire(a.Actor, facing, target, true))
+				if (!a.CheckFire(a.Actor, facing, target))
 					continue;
 
 				if (a.Info.MuzzleSequence != null)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:191]: It looks to me like this here is supposed to simply call all the `notifyAttacks` traits instead of awkwardly adding them to `Armament`. That would be a much cleaner solution. (Even better would be `Armament` offering an event or listener you can add to, but seeing that this already does the muzzle sequence this shouldn't be a problem, so I wouldn't complicate things unless necessary.)

**Review by abcdefg30 — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 3 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs]: this is basically a static list, it will change incredibly rarely and for very few units. I thought this was more appropriate. And by this point to me it feels like normal code rather than gymnastics 😅

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs:360]: it is still within the delayed action

File: OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:191]: > It looks to me like this here is supposed to simply call all the notifyAttacks traits instead of awkwardly adding them to Armament

that's how it used to work, and it looked so similar to other implementations that I missed the difference and accidentally created #21093. The issue is that CheckFire can fire multiple times per tick, so the armament needs to be the one handling the notifying

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Armament.cs]: Maybe it is better to use 
```List<(Actor NotifyActor, INotifyAttack[] Notifies)> notifyAttacks;```, which will save us some perf when adding or removing, as well as faster on searching.

> this is basically a static list, it will change incredibly rarely and for very few units.

Passing `notifyAttacks` between different actors can be a hot function for third party modders in their down stream code, we need some perfs here.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs]: it was needed

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 3 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs]: it wouldn't happen often enough in third part mods as well. I changed it to use a list, and it's not really much cleaner

File: OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs (hunk 4 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:191]: You mean `DoAttack` runs multiple times? I meant running the notifies between the continue and the next if.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 2 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Armament.cs:360]: Ah my bad, the braces are from the if.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Armament.cs]: Hm, why? `notifyAttacks` is the local variable here and not modified, while `this.notifyAttacks` is modified? 🤔

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs]: Just append requires an array type. But we aren't using append anymore anyhow

File: OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:191]: no, the `CheckFire` contains a do-while loop that can fire multiple times

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs:191]: I see. One would assume that it isn't necessary to invoke the notifiers several times in the same tick, but it is inside a delayed action.

**Review by abcdefg30 — APPROVED** [review]
Code changes lgtm, but untested.

P.S. It feels like AttackGarrisoned shouldn't (re-)use Armaments from the passengers, but that is not for this PR.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2797d5e16e2c0077add2438a426a1694ca39bfab...41f5c5df88de009653f024b4685b06b51944b3a5)
